src/compiler/tsc.ts

```diff
@@ -279,9 +279,15 @@ namespace ts {
         }
 
         // If a source file changes, mark it as unwatched and start the recompilation timer
-        function sourceFileChanged(sourceFile: SourceFile) {
+        function sourceFileChanged(sourceFile: SourceFile, removed: boolean) {
             sourceFile.fileWatcher.close();
             sourceFile.fileWatcher = undefined;
+            if (removed){
+                var index = rootFileNames.indexOf(sourceFile.fileName);
+                if (index !== -1){
+                    rootFileNames.splice(index, 1);

```

Comment:
Does this cover referenced files?

Also, I don't know if this is the right approach. I think that if a root file is deleted, we should be removing the file from the cache, but we should be erroring when we fail to find it.